File: app/src/main/java/com/zeyad/rxredux/screens/user/UserDiffCallBack.java
package com.zeyad.rxredux.screens.user;

import android.support.annotation.Nullable;
import android.support.v7.util.DiffUtil;

import com.zeyad.gadapter.ItemInfo;

import java.util.List;

/**
 * @author ZIaDo on 12/13/17.
 */

public class UserDiffCallBack extends DiffUtil.Callback {

    List<ItemInfo> oldUsers;
    List<ItemInfo> newUsers;

    public UserDiffCallBack(List<ItemInfo> newUsers, List<ItemInfo> oldUsers) {
        this.newUsers = newUsers;
        this.oldUsers = oldUsers;
    }

    @Override
    public int getOldListSize() {
        return oldUsers.size();
    }

    @Override
    public int getNewListSize() {
        return newUsers.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return oldUsers.get(oldItemPosition).getId() == newUsers.get(newItemPosition).getId();
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return oldUsers.get(oldItemPosition).<User>getData().equals(newUsers.get(newItemPosition)
                .<User>getData());
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        //you can return particular field for changed item.
        return super.getChangePayload(oldItemPosition, newItemPosition);
    }
}

File: app/src/main/java/com/zeyad/rxredux/screens/user/UserDiffCallBack.kt
package com.zeyad.rxredux.screens.user

import android.support.v7.util.DiffUtil

import com.zeyad.gadapter.ItemInfo

class UserDiffCallBack(private var newUsers: List<ItemInfo>,
                       private var oldUsers: List<ItemInfo>) : DiffUtil.Callback() {

    override fun getOldListSize(): Int {
        return oldUsers.size
    }

    override fun getNewListSize(): Int {
        return newUsers.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldUsers[oldItemPosition].id == newUsers[newItemPosition].id
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldUsers[oldItemPosition].getData<User>() == newUsers[newItemPosition]
                .getData<User>()
    }

    override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
        //you can return particular field for changed item.
        return super.getChangePayload(oldItemPosition, newItemPosition)
    }
}
